### ISSUE TITLE
There is no way to store sessions in the Redis

### ISSUE THREAD

**Issue by samm-git** [magento/magento2#2256]
M2 includes Redis cache session storage from CM repository, but on the same time session storage module is not included. So customer have to use database, memcached or files. Is it expected to be this way? Latest M1 EE includes redis session support.

**Comment by choukalos** [comment]
This worked for me about 6 months ago ... 

Install Redis
Assuming this is already installed or relevant OS specific instructions can be found online.
Install phpredis
https://github.com/nicolasff/phpredis#installation
Update the php.ini file
extension=/path/to/your/redis.so
Restart the web server
Update the app/etc/env.php file to replace the existing <session> with the following:
'session' =>
  array (
    'save' => 'redis',
    'save_path' => 'tcp://127.0.0.1:6379?weight=2&timeout=2.5',
  ),
Be sure to replace the <save_path> information with details needed to connect to your redis server.
That should be all that is needed. You can use the `keys PHPREDIS_SESSION*` command from redis-cli to verify that it is working.

**Comment by samm-git** [comment]
phpredis handler do not support locking, so it is very dangerous and not recommended.

**Comment by choukalos** [comment]
Yes - just realized it's on our roadmap to add locking but not done yet .... preferred method is Memcache....

- PECL memcache extension must be installed
- env.php updated like below: 

'session' =>
  array (
    'save' => 'memcache',
    'save_path' => 'tcp://{HOST}:{PORT}?persistent=1&weight=2&timeout=10&retry_interval=10'
  )

**Comment by samm-git** [comment]
i know about this. just wanted to add that it is regression compared to m1 and would be great to fix this.

**Comment by johnbendi** [comment]
I just upgraded to php7 and getting the following errors:

```
Warning: ini_set(): Cannot find save handler 'memcache' in /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/Session/SessionManager.php on line 535
#0 [internal function]: Magento\Framework\App\ErrorHandler->handler(2, 'ini_set(): Cann...', '/usr/share/ngin...', 535, Array)
#1 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/Session/SessionManager.php(535): ini_set('session.save_ha...', 'memcache')
#2 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/Session/SessionManager.php(184): Magento\Framework\Session\SessionManager->initIniOptions()
#3 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/Session/SessionManager.php(130): Magento\Framework\Session\SessionManager->start()
#4 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/ObjectManager/Factory/AbstractFactory.php(114): Magento\Framework\Session\SessionManager->__construct(Object(Magento\Framework\App\Request\Http), Object(Magento\Framework\Session\SidResolver\Proxy), Object(Magento\Backend\Model\Session\AdminConfig), Object(Magento\Framework\Session\SaveHandler), Object(Magento\Framework\Session\Validator), Object(Magento\Framework\Session\Storage), Object(Magento\Framework\Stdlib\Cookie\PhpCookieManager), Object(Magento\Framework\Stdlib\Cookie\CookieMetadataFactory), Object(Magento\Framework\App\State))
#5 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/ObjectManager/Factory/Dynamic/Developer.php(89): Magento\Framework\ObjectManager\Factory\AbstractFactory->createObject('Magento\\Backend...', Array)
#6 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/ObjectManager/ObjectManager.php(71): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('Magento\\Backend...')
#7 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/ObjectManager/Factory/AbstractFactory.php(236): Magento\Framework\ObjectManager\ObjectManager->get('Magento\\Backend...')
#8 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/ObjectManager/Factory/Dynamic/Developer.php(53): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument(Array, 'Magento\\Framewo...', NULL, 'session', 'Magento\\Store\\A...')
#9 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/ObjectManager/Factory/Dynamic/Developer.php(82): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->_resolveArguments('Magento\\Store\\A...', Array, Array)
#10 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/ObjectManager/ObjectManager.php(71): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('Magento\\Store\\A...')
#11 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/ObjectManager/Factory/AbstractFactory.php(236): Magento\Framework\ObjectManager\ObjectManager->get('Magento\\Store\\A...')
#12 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/ObjectManager/Factory/Dynamic/Developer.php(53): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument(Array, 'Magento\\Framewo...', NULL, 'redirect', 'Magento\\Backend...')
#13 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/ObjectManager/Factory/Dynamic/Developer.php(82): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->_resolveArguments('Magento\\Backend...', Array, Array)
#14 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/ObjectManager/ObjectManager.php(71): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('Magento\\Backend...')
#15 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/ObjectManager/Factory/AbstractFactory.php(236): Magento\Framework\ObjectManager\ObjectManager->get('Magento\\Backend...')
#16 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/ObjectManager/Factory/Dynamic/Developer.php(53): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument(Array, 'Magento\\Backend...', NULL, 'context', 'Magento\\Backend...')
#17 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/ObjectManager/Factory/Dynamic/Developer.php(82): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->_resolveArguments('Magento\\Backend...', Array, Array)
#18 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/ObjectManager/ObjectManager.php(57): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('Magento\\Backend...', Array)
#19 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/App/ActionFactory.php(37): Magento\Framework\ObjectManager\ObjectManager->create('Magento\\Backend...')
#20 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/App/Router/Base.php(300): Magento\Framework\App\ActionFactory->create('Magento\\Backend...')
#21 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/App/Router/Base.php(161): Magento\Framework\App\Router\Base->matchAction(Object(Magento\Framework\App\Request\Http), Array)
#22 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/App/FrontController.php(50): Magento\Framework\App\Router\Base->match(Object(Magento\Framework\App\Request\Http))
#23 [internal function]: Magento\Framework\App\FrontController->dispatch(Object(Magento\Framework\App\Request\Http))
#24 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/Interception/Interceptor.php(74): call_user_func_array(Array, Array)
#25 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/Interception/Chain/Chain.php(70): Magento\Framework\App\FrontController\Interceptor->___callParent('dispatch', Array)
#26 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/Interception/Chain/Chain.php(63): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Framewo...', 'dispatch', Object(Magento\Framework\App\FrontController\Interceptor), Array, 'install')
#27 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/Module/Plugin/DbStatusValidator.php(69): Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}(Object(Magento\Framework\App\Request\Http))
#28 [internal function]: Magento\Framework\Module\Plugin\DbStatusValidator->aroundDispatch(Object(Magento\Framework\App\FrontController\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#29 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/Interception/Chain/Chain.php(67): call_user_func_array(Array, Array)
#30 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/Interception/Interceptor.php(136): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Framewo...', 'dispatch', Object(Magento\Framework\App\FrontController\Interceptor), Array, 'storeCookieVali...')
#31 /usr/share/nginx/html/magedemo.next/app/code/Magento/Store/Model/Plugin/StoreCookie.php(78): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#32 [internal function]: Magento\Store\Model\Plugin\StoreCookie->aroundDispatch(Object(Magento\Framework\App\FrontController\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#33 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/Interception/Interceptor.php(140): call_user_func_array(Array, Array)
#34 /usr/share/nginx/html/magedemo.next/var/generation/Magento/Framework/App/FrontController/Interceptor.php(26): Magento\Framework\App\FrontController\Interceptor->___callPlugins('dispatch', Array, Array)
#35 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/App/Http.php(115): Magento\Framework\App\FrontController\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#36 /usr/share/nginx/html/magedemo.next/lib/internal/Magento/Framework/App/Bootstrap.php(258): Magento\Framework\App\Http->launch()
#37 /usr/share/nginx/html/magedemo.next/pub/index.php(37): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http))
#38 {main}
```

I have `memcached` installed at 11211 and have the following session related snippet in `env.php`:

```
  'session' => array (
    'save' => 'memcache',
    'save_path' => 'tcp://127.0.0.1:11211',
  )
```

**Comment by johnbendi** [comment]
To resolve this I had to install `php-memcached` and configure session in `env.php` thusly:

```
  'session' => array (
    'save' => 'memcached',
    'save_path' => '127.0.0.1:11211',
  )
```

Note that memcached and php7 does not need the `tcp://` prefix.

**Comment by samm-git** [comment]
@johnbendi  hi, this is covered in the magento installation manual.

**Comment by choukalos** [comment]
Closing as this was resolved in 2.0.6 patch & the 2.1.0 release (locking sessions)
